### DISCUSSION TITLE
Map scalebar always horizontal even for projections not preserving angles (Mercator vs. RGF93)

### DISCUSSION THREAD

**Discussion by salahelfarissi** [raphaelquast/EOmaps#251] [Q&A] [answered]
I am using your library to plot maps, currently I'm using Web Mercator as a projection (Below is a map in web mercator).
Recently I was asked to use a more precise projection. For France, we use 2154. However, when I do that the Scalebar is not horizontal anymore. Is there a solution to my problem.

Thank you

![test_carte_mco-extension_e_occupation_du_s_a3](https://github.com/user-attachments/assets/817c482e-fecf-4f64-b6a5-b10e4915c1e3)

**Comment by raphaelquast** [comment]
Hey, nice to see you're using EOmaps! 

the scalebar has a "rotation" argument that sets the azimuth angle in which the scalebar points 
[See docs for add_scalebar](https://eomaps.readthedocs.io/en/latest/generated/eomaps.eomaps.Maps.add_scalebar.html#eomaps.eomaps.Maps.add_scalebar)

As this angle is relative to the direction of latitude lines, a scalebar with `rotation=0` is not necessarily "horizontal" in the plot (except for 4326 and 3857). 

As a quick suggestion i see 2 possible solutions:
- the simple way: manually adjust the angle until the scalebar is approx. horizontal for the desired the position (you can use +- keys to interactively rotate the scalebar and get the active rotation with [scalebar_instance.get_position](https://eomaps.readthedocs.io/en/latest/generated/eomaps.scalebar.ScaleBar.get_position.html#eomaps.scalebar.ScaleBar.get_position))
- the nice but complex way: use `m.ax.projection.get_geod()` to get the `pyproj.geod` instance for the used crs and then use pyproj methods to determine the desired orientation angle for a given position...

**Comment by raphaelquast** [comment]
Hey, I had a proper look at the problem and actually found a quite nice and simple solution!
It will be available as `ScaleBar.set_rotation` with the next release!

In the meantime, here's a quick monkey-patch version:

```python
from eomaps import Maps
import numpy as np

def set_rotation(self, ang=0):
    """
    Set the absolute rotation angle of the first segment of the scalebar.
    
    Note
    ----
    This method sets the "absolute rotation angle" in display units, 
    not the "azimuth angle" which can be set with :py:meth:`ScaleBar.set_position`.
    
    Parameters
    ----------
    ang : float
        The rotation angle.
    
    """
    lon, lat, _ = self.get_position()
    x0, y0 = self._m._transf_lonlat_to_plot.transform(lon, lat)
    x1, y1 = self._m._transf_lonlat_to_plot.transform(
        *self._geod.fwd(lon, lat, -ang, self.get_scale())[:2]
        )
    azim = np.rad2deg(-np.arctan2(x1 - x0, y1 - y0))
    self.set_position((lon, lat), azim=azim)

m = Maps(2154)
m.add_feature.preset.coastline()

scb1 = m.add_scalebar((-6, 43.45))
set_rotation(scb1, 0)

scb2 = m.add_scalebar((2.5, 46))
set_rotation(scb2, 45)

scb3 = m.add_scalebar((-7, 51))
set_rotation(scb3, 90)
```

![grafik](https://github.com/user-attachments/assets/210b4831-6961-490c-9aa4-8402a193f126)

**Comment by salahelfarissi** [comment]
Thank you.
We use your library in a celery context to produce gis maps.
I will test this and give you my feedback.
I hope to be able to contribute to your project very soon.

One last remark that I got from some of our collaborators.
Is there a way to have rounded values, preferable to x5 / x10 / x100 ...

![image](https://github.com/user-attachments/assets/237a87e7-d89a-4ee4-9390-548dbed75161)

Currently I'm planning to do that using the x limits of the map, to specify directly the segment length, I'm obliged to have 4 segments alternating in color.

**Comment by raphaelquast** [comment]
Hey,

>  I hope to be able to contribute to your project very soon.

🚀  Nice to hear! I very much appreciate any efforts to improve the package!


> Is there a way to have rounded values, preferable to x5 / x10 / x100 ...

There is an example in the docs on how to fully customize scalebars: 
[Example - Scalebars](https://eomaps.readthedocs.io/en/latest/EOmaps_examples.html#example-scalebars)

The "scale" (e.g. the length of one segment of the scalebar) is only dependent on the axes-limits if you do not specify it explicitly.
Just use `m.add_scalebar(scale=29)` and you should be good to go!

For example:
```python
from eomaps import Maps
lon, lat = -2.562, 49.4653
extent = (-2.565, -2.554, 49.4642, 49.4682)

m = Maps(2154)
m.add_feature.preset.coastline()
m.set_extent(extent)

scb1 = m.add_scalebar((lon, lat))
scb2 = m.add_scalebar((lon+0.002, lat), scale=29, n=4,  
                      scale_props=dict(colors=("r", "g")))
scb3 = m.add_scalebar((lon+0.004, lat), scale=29, n=4, 
                      scale_props=dict(colors=("k", ".75")), 
                      patch_props=dict(fc="none", ec="none"))
```


![grafik](https://github.com/user-attachments/assets/d496720e-1a71-4e53-a18a-bbab96f51859)

**Comment by raphaelquast** [comment]
Hey, I've converted this to a discussion to preserve the answers for later!
Feel free to mark as accepted if you think all your questions are answered!

**Comment by raphaelquast** [comment, marked as answer]
With an update to [EOmaps v8.3](https://github.com/raphaelquast/EOmaps/releases/tag/v8.3) you can use:

```python
eomaps import Maps

m = Maps(2154)
m.add_feature.preset.coastline()
scb = m.add_scalebar((-6, 43.45), scale=5e4)
scb.set_rotation(90)
```
